crypto/params.c

```diff
@@ -0,0 +1,758 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/params.h>
+#include "internal/thread_once.h"
+
+#define SET_RETURN_SIZE(p, sz) \
+    if ((p)->return_size != NULL) \
+        *(p)->return_size = (sz)
+
+const OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key)
+{
+    if (p != NULL && key != NULL)
+        for (; p->key != NULL; p++)
+            if (strcmp(key, p->key) == 0)
+                return p;
+    return NULL;
+}
+
+static OSSL_PARAM ossl_param_construct(const char *key, unsigned int data_type,
+                                       void *buffer, size_t buffer_size,
+                                       size_t *return_size)
+{
+    OSSL_PARAM res;
+
+    res.key = key;
+    res.data_type = data_type;
+    res.buffer = buffer;
+    res.buffer_size = buffer_size;
+    res.return_size = return_size;
+    return res;
+}
+
+int OSSL_PARAM_get_int(const OSSL_PARAM *p, int *val)
+{
+    if (val == NULL
+        || p == NULL
+        || (p->data_type != OSSL_PARAM_INTEGER
+            && p->data_type != OSSL_PARAM_UNSIGNED_INTEGER))
+         return 0;
+    switch (p->buffer_size) {
+    case sizeof(int32_t):
+        if (sizeof(int) >= sizeof(int32_t)) {
+            *val = (int)*(const int32_t *)p->buffer;
+            return 1;
+        }
+        break;
+    case sizeof(int64_t):
+        if (sizeof(int) >= sizeof(int64_t)) {
+            *val = (int)*(const int64_t *)p->buffer;
+            return 1;
+        }
+        break;
+    }
+    return 0;
+}
+
+int OSSL_PARAM_set_int(const OSSL_PARAM *p, int val)
+{
+    if (p == NULL)
+        return 0;
```

Comment:
We should fail if `p->data_type == OSSL_PARAM_UNSIGNED_INTEGER` and `val < 0`

Reply:
Depends.  Do we want to forbit signed -> unsigned casts?  That's what you're implying.

Reply:
I think if the OSSL_PARAM structure has been set up to expect an unsigned integer and we try to set a negative value then it should just fail. If you really want to cast -ve values to their unsigned equivalent then you should call `OSSL_PARAM_set_uint`

Reply:
The simple fix is only allow `OSSL_PARAM_UNSIGNED_INTEGER` with `uint`

Reply:
It’s an easy enough check and does catch a known exploit scenario. Likewise, a too large unsigned to signed conversion could return error as mooted previously.

Reply:
Similarly I'm not seeing a change as a result of this?